emily/cdk/lib/emily-stack.ts

```diff
@@ -0,0 +1,196 @@
+
+/**
+* @class CloudFormationStack
+* @classdesc Creates a stack with DynamoDB tables and a Lambda function.
+*/
+export class EmilyStack extends cdk.Stack {
+
+    /**
+    * @constructor
+    * @param {Construct} scope The AWS CDK construct scope.
+    * @param {string} id The stack ID.
+    * @param {EmilyStackProps} props The stack properties.
+    */
+    constructor(scope: Construct, id: string, props: EmilyStackProps) {
+        super(scope, id, props);
+        const depositTable: dynamodb.Table = this.createOrUpdateDepositTable(props);
+        const withdrawalTable: dynamodb.Table = this.createOrUpdateWithdrawalTable(props);
+        const chainstateTable: dynamodb.Table = this.createOrUpdateChainstateTable(props);
+        const operationLambda: lambda.Function = this.createOrUpdateOperationLambda(
+            depositTable,
+            withdrawalTable,
+            chainstateTable,
+            props
+        );
+        const emilyApi: apig.SpecRestApi = this.createOrUpdateApi(operationLambda, props);
+    }
+
+    createOrUpdateDepositTable(props: EmilyStackProps): dynamodb.Table {
+        const tableId: string = 'DepositTable';
+        const table: dynamodb.Table = new dynamodb.Table(this, tableId, {
+            tableName: EmilyStackUtils.getResourceName(tableId, props),
+            partitionKey: {
+                name: 'BitcoinTxid',
+                type: dynamodb.AttributeType.BINARY,
+            },
+            sortKey: {
+                name: 'BitcoinTxOutputIndex',
+                type: dynamodb.AttributeType.NUMBER,
+            }
+        });
+
+        const indexName: string = "DepositStatus";
+        table.addGlobalSecondaryIndex({
+            indexName: indexName,
+            partitionKey: {
+                name: 'OpStatus',
+                type:  dynamodb.AttributeType.NUMBER
+            },
+            sortKey: {
+                name: 'LastUpdateHeight',
+                type:  dynamodb.AttributeType.NUMBER
+            },
+            projectionType: dynamodb.ProjectionType.INCLUDE,
+            nonKeyAttributes: [
+                "BitcoinTxid",
+                "BitcoinTxOutputIndex",
+                "Recipient",
+                "Amount",
+            ]
+        });
+
+        // TODO: Add an additional GSI for querying by user; not required for MVP.
+        return table;
+    }
+
+    createOrUpdateWithdrawalTable(props: EmilyStackProps): dynamodb.Table {
+        // Create DynamoDB table to store the messages. Encrypted by default.
+        const tableId: string = 'WithdrawalTable';
+        const table: dynamodb.Table = new dynamodb.Table(this, tableId, {
+            tableName: EmilyStackUtils.getResourceName(tableId, props),
+            partitionKey: {
+                name: 'RequestId',
+                type: dynamodb.AttributeType.STRING,
+            },
+            sortKey: {
+                name: 'StacksBlockHash',
+                type: dynamodb.AttributeType.BINARY,
```

Comment:
The `RequestId` is _not_ globally unique, but just unique per stacks block hash? Also, why is the `StacksBlockHash` not a string? Not saying any of this is wrong it was just surprising so I figured I'd ask.

Reply:
The `RequestId` should be unique, but there's a brief moment on a fork when it isn't, so this avoids the collision without needing any extra forking related info that's unknown at creation.

Reply:
RequestId should be a number though, oops

Reply:
All of this is good to know, I'm glad I asked.